### PR TITLE
move kube proxy config to config object

### PR DESCRIPTION
**Pull request by mikedanese**

### CODE REVIEW

File: cmd/kube-proxy/app/server.go [hunk 1]
```diff
@@ -201,9 +193,9 @@ func NewProxyServerDefault(config *ProxyServerConfig) (*ProxyServer, error) {
 
 	// TODO(vmarmol): Use container config for this.
 	var oomAdjuster *oom.OOMAdjuster
-	if config.OOMScoreAdj != 0 {
+	if *config.OOMScoreAdj != 0 {
```
Comment (ArtfulCoder) [inline cmd/kube-proxy/app/server.go]: shoudl we do null checks ?

File: cmd/kube-proxy/app/server.go [hunk 2]
```diff
@@ -123,19 +114,20 @@ func checkKnownProxyMode(proxyMode string) bool {
...
-		OOMScoreAdj:                    qos.KubeProxyOOMScoreAdj,
-		ResourceContainer:              "/kube-proxy",
-		IptablesSyncPeriod:             30 * time.Second,
-		ConfigSyncPeriod:               15 * time.Minute,
-		KubeAPIQPS:                     5.0,
-		KubeAPIBurst:                   10,
-		UDPIdleTimeout:                 250 * time.Millisecond,
-		ConntrackMax:                   256 * 1024, // 4x default (64k)
-		ConntrackTCPTimeoutEstablished: 86400,      // 1 day (1/5 default)
+		KubeProxyConfiguration: componentconfig.KubeProxyConfiguration{
```
Comment (ArtfulCoder) [inline cmd/kube-proxy/app/server.go]: ensure that defaults dont change. some are...

File: cmd/kube-proxy/app/server.go (hunk 1 above)
Comment (smarterclayton) [inline cmd/kube-proxy/app/server.go]: Yes

File: cmd/kube-proxy/app/server.go [hunk 3]
```diff
@@ -85,26 +76,26 @@ type ProxyServer struct {
...
-	fs.StringVar(&s.ProxyMode, "proxy-mode", "", "Which proxy mode to use: 'userspace' (older) or 'iptables' (faster). If blank, look at the Node object on the Kubernetes API and respect the '"+experimentalProxyModeAnnotation+"' annotation if provided.  Otherwise use the best-available proxy (currently iptables).  If the iptables proxy is selected, regardless of how, but the system's kernel or iptables versions are insufficient, this always falls back to the userspace proxy.")
-	fs.DurationVar(&s.IptablesSyncPeriod, "iptables-sync-period", s.IptablesSyncPeriod, "How often iptables rules are refreshed (e.g. '5s', '1m', '2h22m').  Must be greater than 0.")
+	//fs.StringVar(&s.Mode, "proxy-mode", s.Mode, "Which proxy mode to use: 'userspace' (older) or 'iptables' (faster). If blank, look at the Node object on the Kubernetes API and respect the '"+experimentalProxyModeAnnotation+"' annotation if provided.  Otherwise use the best-available proxy (currently iptables).  If the iptables proxy is selected, regardless of how, but the system's kernel or iptables versions are insufficient, this always falls back to the userspace proxy.")
```
Comment (mikedanese) [inline cmd/kube-proxy/app/server.go]: fix this up

Comment (smarterclayton) [inline cmd/kube-proxy/app/server.go]: Weird - that only appears in this diff - not in my changes.

File: cmd/kube-proxy/app/server.go (hunk 1 above)
Comment (smarterclayton) [inline cmd/kube-proxy/app/server.go]: Hrm this is some kind of bizarre reverse diff...

File: cmd/kube-proxy/app/server.go (hunk 3 above)
Comment (smarterclayton) [inline cmd/kube-proxy/app/server.go]: Oh wow, wrong pr.

File: cmd/kube-proxy/app/server.go (hunk 1 above)
Comment (smarterclayton) [inline cmd/kube-proxy/app/server.go]: Ignore, GitHub tricked me.

**Comment by mikedanese** [comment]
@ArtfulCoder Done PTAL!

File: cmd/kube-proxy/app/options/options.go [hunk 4]
```diff
@@ -33,67 +34,59 @@ const (
...
+		KubeAPIQPS:       5.0,
+		KubeAPIBurst:     10,
+		ConfigSyncPeriod: 15 * time.Minute,
 	}
 }
 
 // AddFlags adds flags for a specific ProxyServer to the specified FlagSet
 func (s *ProxyServerConfig) AddFlags(fs *pflag.FlagSet) {
-	fs.IPVar(&s.BindAddress, "bind-address", s.BindAddress, "The IP address for the proxy server to serve on (set to 0.0.0.0 for all interfaces)")
+	fs.StringVar(&s.BindAddress, "bind-address", s.BindAddress, "The IP address for the proxy server to serve on (set to 0.0.0.0 for all interfaces)")
```
Comment (ArtfulCoder) [inline cmd/kube-proxy/app/options/options.go]: I dont follow why we need BindAddress to be a string.
Can it not stay as IP ?

Comment (mikedanese) [inline cmd/kube-proxy/app/options/options.go]: It cannot stay an IP because we don't have a representation of IP in the API.

Comment (mikedanese) [inline cmd/kube-proxy/app/options/options.go]: I tried to do this but Tim said it was a bad idea. https://github.com/kubernetes/kubernetes/pull/18689#issuecomment-164644315

File: cmd/kube-proxy/app/options/options.go [hunk 5]
```diff
@@ -33,67 +34,59 @@ const (
 	ExperimentalProxyModeAnnotation = "net.experimental.kubernetes.io/proxy-mode"
 )
 
-// ProxyServerConfig contains configurations for a Kubernetes proxy server
+// ProxyServerConfig contains configures and runs a Kubernetes proxy server
```
Comment (ArtfulCoder) [inline cmd/kube-proxy/app/options/options.go]: nit: "configures" ?

File: cmd/kube-proxy/app/server.go [hunk 6]
```diff
@@ -202,7 +204,14 @@ func NewProxyServerDefault(config *options.ProxyServerConfig) (*ProxyServer, err
 		// set EndpointsConfigHandler to our loadBalancer
 		endpointsHandler = loadBalancer
 
-		proxierUserspace, err := userspace.NewProxier(loadBalancer, config.BindAddress, iptInterface, config.PortRange, config.IptablesSyncPeriod, config.UDPIdleTimeout)
+		proxierUserspace, err := userspace.NewProxier(
+			loadBalancer,
+			net.ParseIP(config.BindAddress),
+			iptInterface,
+			*utilnet.ParsePortRangeOrDie(config.PortRange),
```
Comment (ArtfulCoder) [inline cmd/kube-proxy/app/server.go]: can we validate during flag parsing ?

File: pkg/apis/componentconfig/helpers.go [hunk 7]
```diff
@@ -48,3 +50,18 @@ func (v IPVar) String() string {
...
+func (m *ProxyMode) Set(s string) error {
+	nm := ProxyMode(s)
+	m = &nm
+	return nil
+}
+func (m *ProxyMode) String() string {
+	if m != nil {
+		return string(*m)
+	}
+	return "<nil>"
```
Comment (ArtfulCoder) [inline pkg/apis/componentconfig/helpers.go]: empty string ?

**Comment by ArtfulCoder** [comment]
Looking good. minor comments need addressing..

File: cmd/kube-proxy/app/server.go (hunk 6 above)
Comment (mikedanese) [inline cmd/kube-proxy/app/server.go]: DOne

**Comment by mikedanese** [comment]
@k8s-bot ok to test

**Comment by mikedanese** [comment]
@k8s-bot unit test this please

**Comment by mikedanese** [comment]
```
14:54:43 Verifying ./hack/../hack/verify-codegen.sh
14:54:43 +++ [0128 22:54:43] Building go targets for linux/amd64:
14:54:43     cmd/libs/go2idl/client-gen
14:54:48 +++ [0128 22:54:48] Placing binaries
14:54:49 +++ [0128 22:54:49] Building go targets for linux/amd64:
14:54:49     cmd/libs/go2idl/deepcopy-gen
14:54:51 +++ [0128 22:54:51] Placing binaries
14:54:52 +++ [0128 22:54:52] Building go targets for linux/amd64:
14:54:52     cmd/libs/go2idl/set-gen
14:54:54 +++ [0128 22:54:54] Placing binaries
14:54:56 type checker error: github.com/opencontainers/runc/libcontainer/cgroups/fs/cpuacct.go:22:25: GetClockTicks not declared by package system
14:54:56 type checking encountered some errors in "github.com/opencontainers/runc/libcontainer/cgroups/fs", but ignoring.
14:54:57 2016/01/28 22:54:57 Processing package "release_1_1", disk location "/workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/client/clientset_generated/release_1_1"
14:54:57 2016/01/28 22:54:57 Verifying file "/workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/client/clientset_generated/release_1_1/clientset.go"
14:54:57 2016/01/28 22:54:57 Processing package "unversioned", disk location "/workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/client/typed/generated/legacy/unversioned"
14:54:57 2016/01/28 22:54:57 Verifying file "/workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/client/typed/generated/legacy/unversioned/doc.go"
14:54:57 2016/01/28 22:54:57 Verifying file "/workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/client/typed/generated/legacy/unversioned/limitrange.go"
14:54:57 2016/01/28 22:54:57 Verifying file "/workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/client/typed/generated/legacy/unversioned/persistentvolumeclaim.go"
```

@k8s-bot unit test this please

**Comment by matthewdupre** [comment]
@mikedanese I've been working on a pull request (#20496) that adds a configuration option to the kube-proxy, but I've hit problems that seem to have been introduced by this.

I'm testing by using hack/local-up-cluster.sh, and then killing the kube-proxy process and rerunning it with variations on the command line invocation.

I've hit 2 issues:

1. The default for --proxy-mode appears to have been changed from iptables to userspace.  Previously (as of d3fdc61) when the option wasn't specified, you got the iptables proxy, now it's userspace.

2. I'm unable to change the mode at the command line (is this a 'feature'?): --proxy-mode is ignored.  It still appears in the help text however.  I'm going to try the net.experimental.kubernetes.io/proxy-mode annotation.

**Comment by matthewdupre** [comment]
OK apparently 1. is being fixed.

Looks like the annotation works for 2..  The help text is still misleading, but I can at least continue with my PR.

**Comment by mikedanese** [comment]
Thanks for the report. I think this is isolated to the --proxy-mode flag. #20515 should fix

**Pull request merged**